src/create_venv.sh

```diff
@@ -1,136 +1,257 @@
 #!/usr/bin/env bash
 
+# Copyright (c) YugaByte, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+# in compliance with the License.  You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software distributed under the License
+# is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+# or implied.  See the License for the specific language governing permissions and limitations
+# under the License.
+#
 
-# Use SHA of requirements.txt as a comment in requirements_frozen.txt
-# 
+[[ "${_YB_CREATE_VENV_INCLUDED:=""}" == "yes" ]] && return 0
+_YB_CREATE_VENV_INCLUDED=yes
+YB_VERBOSE=${YB_VERBOSE:-false}
+
+set -e -u -o pipefail
+
+_src_dir="${BASH_SOURCE%/*}"
+if [[ ! -d "${_src_dir}" ]]; then _src_dir="$PWD"; fi
+# shellcheck disable=SC1091,SC1090
+. "${_src_dir}"/logger.sh
+# shellcheck disable=SC1091,SC1090
+. "${_src_dir}"/os.sh
+# shellcheck disable=SC1091,SC1090
+. "${_src_dir}"/detect_python.sh
+
+# -------------------------------------------------------------------------------------------------
+# Global variables used in this module
+# -------------------------------------------------------------------------------------------------
+# This preserves the current existing behavior of assuming requirements_frozen.txt is always
+# up to date.  TODO: default this to true to ensure frozen file is 'correct'
+YB_BUILD_STRICT=${YB_BUILD_STRICT:-false}
+# Set this to true to force recreation of requirements_frozen.txt
+YB_RECREATE_VIRTUALENV=${YB_RECREATE_VIRTUALENV:-false}
+# New-style VENV base dir, only used when YB_USE_TOP_LEVEL_VENV is false
+YB_VENV_BASE_DIR=${YB_VENV_BASE_DIR:-~/.venv/yb}
+# This preserves the current existing behavior of putting the VENV in the same dirextory
+# as requirements.txt
+YB_USE_TOP_LEVEL_VENV=${YB_USE_TOP_LEVEL_VENV:-true}
+
+verbose "Using YB_PYTHON_VERSION=${YB_PYTHON_VERSION}"
+# shellcheck disable=SC2154
+verbose "Using ${yb_python_interpreter} (${yb_python_version_actual})"
+
+# -------------------------------------------------------------------------------------------------
+# Internal functions used in this module.  These shouldn't be called directly outside this module.
+# -------------------------------------------------------------------------------------------------
 
 function text_file_sha() {
-  local file="${1}"
+  local file=$1
   local tmp
   tmp="$(sort -u <<<"$(grep -v '^#' "${file}")")"
```

Comment:
We should probably also ignore lines that have whitespace before the `#` character, e.g. with `grep -Ev '^[[:space:]]*#' "$file"`.

We probably also don't need double quotes around the string that is being redirected to `sort -u`.
And we don't need the double quotes around the whole expression. And the curly braces in `${file}`.
In other words, `tmp=$(sort -u <<<$(grep -v '^#' "$file"))` seems more readable to me.

Also is there a benefit for using string redirection instead of a pipe? I think this code is a bit easier to read (not sure about error handling / reliability though):

```
tmp=$( grep -Ev '...' "$file" | sort -u )
```

One difference is that with your approach, `grep` is evaluated first and only then its output is fed to `sort -u`, and with a pipe both commands run concurrently.

Reply:
I think another difference is it doesn't require `-o pipefail` in the subshell.